SCons/Scanner/C.py

```diff
@@ -66,31 +67,69 @@ def read_file(self, file) -> str:
             return ''
 
 def dictify_CPPDEFINES(env) -> dict:
-    """Returns CPPDEFINES converted to a dict.
-
-    This should be similar to :func:`~SCons.Defaults.processDefines`.
-    Unfortunately, we can't do the simple thing of calling that routine and
-    passing the result to the dict() constructor, because it turns the defines
-    into a list of "name=value" pairs, which the dict constructor won't
-    consume correctly.  Also cannot just call dict on CPPDEFINES itself - it's
-    fine if it's stored in the converted form (currently deque of tuples), but
-    CPPDEFINES could be in other formats too.
-
-    So we have to do all the work here - keep concepts in sync with
-    ``processDefines``.
+    """Return CPPDEFINES converted to a dict for preprocessor emulation.
+
+    The concept is similar to :func:`~SCons.Defaults.processDefines`:
+    turn the values stored in an internal form in ``env['CPPDEFINES']``
+    into one usable in a specific context - in this case the cpp-like
+    work the C/C++ scanner will do. We can't reuse ``processDefines``
+    output as that's a list of strings for the command line. We also can't
+    pass the ``CPPDEFINES`` variable directly to the ``dict`` constructor,
+    as SCons allows it to be stored in several different ways - it's only
+    after ``Append`` and relatives has been called we know for sure it will
+    be a deque of tuples.
+
+    Since the result here won't pass through a real preprocessor, simulate
+    some of the macro replacement that would take place if it did, or some
+    conditional inclusions might come out wrong. A bit of an edge case, but
+    does happen (GH #4623). See 6.10.5 in the C standard and 15.6 in the
+    C++ standard).
+
+    .. versionchanged:: NEXT_RELEASE
+       Simple macro replacement added.
     """
+    def _replace(mapping: Dict) -> Dict:
+        """Simplistic macro replacer for dictify_CPPDEFINES.
+
+        *mapping* is scanned for any value that is the same as a key in
+        the dict, and is replaced by the value of that key; the process
+        is repeated. This is a cheap approximation of the C preprocessor's
+        macro replacement rules with no smarts - it doesn't "look inside"
+        the values, so only triggers on object-like macros, not on
+        function-like macros, and will not work on complex values,
+        e.g. a value like ``(1UL << PR_MTE_TCF_SHIFT)`` would not have
+        ``PR_MTE_TCF_SHIFT`` replaced if it was also in ``CPPDEFINES``,
+        but rather left as-is for the scanner to do comparisons against.
+
+        Args:
+            mapping: a dictionary representing macro names and replacements.
+
+        Returns:
+            a dictionary with substitutions made.
+        """
+        old_ns = mapping
+        ns = {}
+        while True:
+            ns = {k: old_ns[v] if v in old_ns else v for k, v in old_ns.items()}
```

Comment:
This will loop forever if one variable self references. 

`-DXYZ=XYZ`

Reply:
Not convinced... if you have `{'XYZ': 'XYZ'}`, then `k` gets the looked-up value `old_ns['XYZ']`, which is `XYZ`, which is actually the same as it was before. Thus, the following check finds that `old_ns` and `ns` have the same values, so it breaks out of the loop.  There might still be a way to get into a pathological state, but I don't think that example is enough.

Reply:
Perhaps if you have three defines that circularly point to each other?  Is that worth worrying about? cpp does detect that as a pathological case and errors out.

Reply:
Add this to line 581 in CTests.py and run..
```
                    ('XYZ','XYZ')
```

Reply:
Well, of course it's bad, but who would do something like that?  I think fairly simple measures should be enough, as noted above. It's not supposed to be a perfect tool, as the C scanner has a limited job - unlike subst, which has to come out just right.

Reply:
For example - no-loop, just two-pass:
```py
        ns = {k: old_ns[v] if v in old_ns else v for k, v in old_ns.items()}
        if old_ns == ns:  # there were no substitutions
            return ns
        # do one more round
        return {k: ns[v] if v in ns else v for k, v in ns.items()}
```